ayeletK/assignment4/grid.cpp

```diff
@@ -3,59 +3,52 @@
 #include <unsorted_set>
 #include <stdlib.h>
 
-#define DELTA {-1,0,1}
 using namespace std;
 
+vector<Position> delta_positions = {Position(-1,0), Position(0,-1), Position(1,0), Position(0,1)};
+
 template <typename T>
 class Grid{
     int _rows, _cols;
     vector<vector<T>> _grid;
+    
     bool is_legal_row(int row){
         return row >=0 && row < _rows;
     }
     bool is_legal_col(int col){
         return col >=0 && col < _cols;
     }
 public:
-    Grid(): _rows(-1), _cols(-1) {}
-    Grid(int rows_num, int cols_num, bool** data){
-        _rows = rows_num;
-        _cols = cols_num;
-        for (int row = 0; row < _rows; ++row){
-            _grid.push_back(vector<T>(_cols));
-            for (int col = 0; col < _cols; ++col){
-                _grid[row][col] = data[row][col];
-            }
+    Grid(): Grid(nullptr) {}
+    explicit Grid(vector<vector<T>>* data): _rows(-1), _cols(-1) {
+        if (data != nullptr) {
+            _grid = *data;
+            _rows = _grid.size();
+            if (_rows > 0) _cols = _grid[0].size();
         }
     }
-    explicit Grid(vector<vector<T>>* data): _grid(*data) {}
-    void update_value(Position* pos, T val){
-        if (is_legal_position(pos)) 
-            _grid[pos->get_row()][pos->get_col()] = val;
-    }
-    T get_value(Position* pos){
+    T get_value(const Position& pos){
         if (is_legal_position(pos)){
-            return _grid[pos->get_row()][pos->get_col()];
+            return _grid[pos.get_row()][pos.get_col()];
         }
         return T();
     }
     void get_grid_dimentions(int* rows, int* cols){
         *rows = _rows;
         *cols = _cols;
     }
-    bool is_legal_position(Position* pos){
-        return is_legal_row(pos->get_row()) && is_legal_col(pos->get_col());
+    bool is_legal_position(const Position& pos){
+        return is_legal_row(pos.get_row()) && is_legal_col(pos.get_col());
     }
-    unordered_set<Position> get_neighbours(Position* pos, T wanted_val){
+    unordered_set<Position> get_neighbours_of_val(const Position& curr_pos, T wanted_val){
         unordered_set<Position> neighbours;
         Position neighbour;
-        for (int i: DELTA){
-            for (int j: DELTA){
-                if (abs(i-j) != 1) continue;
-                neighbour = Position(pos->get_row()+i, pos->get_col()+j);
-                if (is_legal_position(&neighbour) && get_value(&neighbour) == wanted_val)
-                        neighbours.insert(neighbour);
-            }
+        int row, col;
+        for (Position pos: delta_positions){
+            neighbour = curr_pos + pos;
+            if (neighbour == curr_pos) continue;
```

Comment:
Why would the neighbor be the same as the current position? Does this solve a real problem or is it just for safety?

Industry secret:
If that's just for safety (and you do not expect it to actually happen in runtime), you can replace this with an assert clause:
`assert(neighbor != current_pos);`
This way if this bug happens then this will not fail silently but crash the program so you can find it and debug it.

If you are not worried that it will actually happen, it is best to remove it completely to make the code cleaner.